pulser-simulation/pulser_simulation/simulation.py

```diff
@@ -414,116 +442,51 @@ def draw(
-                for slot in self._seq._schedule[channel]:
-                    if isinstance(slot.type, Pulse):
-                        # If SLM is on during slot, populate local samples
-                        if slm_on and self._seq._slm_mask_time[1] > slot.ti:
-                            samples_dict = self.samples["Local"][basis]
-                            for qubit in slot.targets:
-                                if qubit not in samples_dict:
-                                    samples_dict[qubit] = prepare_dict()
-                                write_samples(
-                                    slot, samples_dict[qubit], True, qubit
-                                )
-                            self.samples["Local"][basis] = samples_dict
-                        # Otherwise, populate corresponding global
-                        else:
-                            slm_on = False
-                            samples_dict = self.samples["Global"][basis]
-                            if not samples_dict:
-                                samples_dict = prepare_dict()
-                            write_samples(slot, samples_dict, True)
-                            self.samples["Global"][basis] = samples_dict
-
-            # Any noise : global becomes local for each qubit in the reg
-            # Since coefficients are modified locally by all noises
-            else:
-                is_global = addr == "Global"
-                samples_dict = self.samples["Local"][basis]
-                for slot in self._seq._schedule[channel]:
-                    if isinstance(slot.type, Pulse):
-                        for qubit in slot.targets:
-                            if qubit not in samples_dict:
-                                samples_dict[qubit] = prepare_dict()
-                            # We don't write samples for badly prep qubits
-                            if not self._bad_atoms[qubit]:
-                                write_samples(
-                                    slot, samples_dict[qubit], is_global, qubit
-                                )
-                self.samples["Local"][basis] = samples_dict
-
-            # Apply SLM mask if it was defined
-            if self._seq._slm_mask_targets and self._seq._slm_mask_time:
-                tf = self._seq._slm_mask_time[1]
-                for qubit in self._seq._slm_mask_targets:
-                    if qubit not in self.samples["Local"][basis]:
-                        continue
-                    for x in ("amp", "det", "phase"):
-                        self.samples["Local"][basis][qubit][x][0:tf] = 0
+            for qid in slot.targets:
+                if "doppler" in self.config.noise:
+                    noise_det = self._doppler_detune[qid]
+                    samples_dict[qid]["det"][slot.ti : slot.tf] += noise_det
+                # Gaussian beam loss in amplitude for global pulses only
+                # Noise is drawn at random for each pulse
```

Comment:
I considered it, but it was not being done before so I think it's out of scope here. 
The sampling being done per pulse rather than per qubit is already going a bit too far, but it didn't make sense to do it the old way and write tests for it that would just have to be rewritten later.
In any case, you can disable one or the other by changing the `laser_waist` and `amp_sigma` parameters.

Reply:
Ok make sense